### DISCUSSION TITLE
Getting started hacking on Z2M - npm commands needed

### DISCUSSION THREAD

**Discussion by lbt** [Koenkk/zigbee2mqtt#10879] [General]
I'm a reasonably competent hacker/developer who really has never learned the whole js/ts/npm/node build and development environment.

I'd like to hack on Z2M and I'd like to be able to run with my changes.

I have been searching around the web but much of the information is years old and it seems to make sense to ask how the devs in this project would suggest I do it.

FWIW I want the Z2M Dashboard page component to show the force() attribute of my TS0601 radiator valve. From code inspection I'm guessing that I need to re-arrange the attributes in the correct fingerprint/manufacturerName in tuya.js (zigbee-herdsman-converters) and *maybe* then add something if the logic in the DashboardFeatureWrapper.tsx (zigbee2mqtt-frontend) doesn't pick up the class... or something :rofl:

The [getting started guide](https://www.zigbee2mqtt.io/guide/installation/01_linux.html#for-later-update-zigbee2mqtt-to-the-latest-version) allows me to hack on z2m code in lib/ but isn't so good on hacking on either zigbee-herdsman-converters or zigbee2mqtt-frontend.

Simply editing /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices/tuya.js and restarting the systemd service didn't work. At least I think it doesn't because I changed the description: to "Radiator valve with *a* thermostat" and the About page for my components didn't change :/ (or is there a cache or ??)

I tried hacking the package.json to point to file:/path-to-git-clone/zigbee-herdsman-converters but that didn't work (even checked out to the same tag as package.json) because:

- `npm ci` complained about "`cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync.`"

- `npm install` complained about "`npm WARN tsutils@3.21.0 requires a peer of typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta but none is installed. You must install peer dependencies yourself.`"

and

```
28274 error code ENOENT
28275 error syscall lstat
28276 error path /opt/zigbee2mqtt/node_modules/.staging/base64-js-08c837c8
28277 error errno -2
28278 error enoent ENOENT: no such file or directory, lstat '/opt/zigbee2mqtt/node_modules/.staging/base64-js-08c837c8'
28279 error enoent This is related to npm not being able to find a file.
28280 verbose exit [ -2, true ]

```

So what I'd ideally like to do is have /opt/zigbee2mqtt contain my 'production' checkout as per the guide and edit package.json to point to
```
     "zigbee-herdsman-converters": "file:/everything/devel/raspi/zigbee-herdsman-converters",
     "zigbee2mqtt-frontend": "file:/everything/devel/raspi/zigbee2mqtt-frontend"
```  
and then know what npm commands to invoke to make it run ;)

**Comment by lbt** [comment]
Any suggestions ? :)

**Reply by Koenkk** [reply]
> Simply editing /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices/tuya.js and restarting the systemd service didn't work. At least I think it doesn't because I changed the description: to "Radiator valve with a thermostat" and the About page for my components didn't change :/ (or is there a cache or ??)

This should work actually, edit -> restart z2m -> reload frontend should do it.
